File: piccolo/apps/api/src/lib/discord.ts
import { env } from "../env.js";
import { logger } from "./logger.js";

/**
 * Ported from notify/discord.py in SimsForTT/tenders (master branch) -
 * that alerting was already built and proven (a real Discord incoming
 * webhook, free, no bot/OAuth setup). This keeps the same message shape
 * and the same graceful "skip if unset" behavior, adapted to Piccolo's
 * tender shape and called from routes/internal.ts instead of a
 * standalone script's main loop.
 */

// Discord hard-caps a single message's `content` at 2000 characters.
const MAX_MESSAGE_LENGTH = 2000;

export type AlertTender = {
  title: string;
  tenderNo?: string | null;
  closingAt?: string | null;
  portal?: string | null;
  url?: string | null;
};

function formatTender(t: AlertTender): string {
  const lines = [`\u{1F195} **${t.title || "Untitled tender"}**`];
  if (t.tenderNo) lines.push(`Ref: ${t.tenderNo}`);
  if (t.closingAt) lines.push(`Closes: ${t.closingAt}`);
  if (t.portal) lines.push(`Portal: ${t.portal}`);
  if (t.url) lines.push(t.url);
  return lines.join("\n");
}

async function sendMessage(content: string): Promise<boolean> {
  if (!env.DISCORD_WEBHOOK_URL) {
    logger.debug("Discord alert skipped: DISCORD_WEBHOOK_URL not set");
    return false;
  }
  try {
    // Discord's webhook API expects a JSON body with a "content" key -
    // Slack/Teams incoming webhooks use "text" instead, which is an easy
    // mismatch to introduce if you swap providers later without updating
    // the field name (see workflows/n8n/vault-expiry-watcher.json).
    const res = await fetch(env.DISCORD_WEBHOOK_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ content }),
    });
    if (!res.ok) {
      logger.warn({ status: res.status }, "Discord alert send failed");
      return false;
    }
    return true;
  } catch (err) {
    logger.warn({ err }, "Discord alert send failed");
    return false;
  }
}

export async function sendNewTenderAlert(tender: AlertTender): Promise<void> {
  const content = formatTender(tender);
  // Discord rejects an over-length message outright rather than
  // truncating it itself - better a clipped alert than a silently
  // dropped one.
  await sendMessage(content.length > MAX_MESSAGE_LENGTH ? content.slice(0, MAX_MESSAGE_LENGTH - 1) + "…" : content);
}

File: notify/discord.py
"""
Phase 3: Discord alerting via an incoming webhook (free, no bot required).

Setup (one-time, on your machine):
    1. In your Discord server: Server Settings -> Integrations -> Webhooks
       -> New Webhook. Pick the channel you want alerts posted to.
    2. Copy the webhook URL.
    3. Put it in .env (copy from .env.example):
       DISCORD_WEBHOOK_URL=...

Usage:
    from notify.discord import send_alert
    send_alert(tender_dict)
    # or for a batch:
    send_alert_batch(list_of_tender_dicts)
"""

import os

import httpx
from dotenv import load_dotenv

load_dotenv()

DISCORD_WEBHOOK_URL = os.getenv("DISCORD_WEBHOOK_URL")

# Discord hard-caps a single message's `content` at 2000 characters.
MAX_MESSAGE_LENGTH = 2000


def _format_tender(tender: dict) -> str:
    lines = [f"🆕 **{tender.get('title', 'Untitled tender')}**"]
    if tender.get("ref_number"):
        lines.append(f"Ref: {tender['ref_number']}")
    if tender.get("closing_date"):
        lines.append(f"Closes: {tender['closing_date']}")
    if tender.get("portal"):
        lines.append(f"Portal: {tender['portal']}")
    if tender.get("url"):
        lines.append(tender["url"])
    return "\n".join(lines)


def _chunk_blocks(header: str, blocks: list[str]) -> list[str]:
    """Pack blocks into messages, each kept under MAX_MESSAGE_LENGTH."""
    chunks = []
    current = header
    for block in blocks:
        candidate = f"{current}\n\n{block}" if current else block
        if len(candidate) > MAX_MESSAGE_LENGTH:
            if current:
                chunks.append(current)
            current = block
        else:
            current = candidate
    if current:
        chunks.append(current)
    return chunks


def _send_message(text: str) -> bool:
    if not DISCORD_WEBHOOK_URL:
        print("  [discord] Skipped: DISCORD_WEBHOOK_URL not set in .env")
        return False

    try:
        resp = httpx.post(DISCORD_WEBHOOK_URL, json={"content": text}, timeout=10)
        resp.raise_for_status()
        return True
    except Exception as e:
        print(f"  [discord] Send failed: {e}")
        return False


def send_alert(tender: dict) -> bool:
    """Send a single tender alert. Returns True if sent successfully."""
    return _send_message(_format_tender(tender))


def send_alert_batch(tenders: list[dict]) -> bool:
    """Send one or more messages summarizing multiple new tenders."""
    if not tenders:
        return True
    header = f"📋 {len(tenders)} new tender(s) found:"
    blocks = [_format_tender(t) for t in tenders]
    chunks = _chunk_blocks(header, blocks)
    return all(_send_message(chunk) for chunk in chunks)
